website/src/components/FlaggableElement.tsx

```diff
@@ -1,152 +1,84 @@
-}
-
-interface LabelValue {
-  label: Label;
-  checked: boolean;
-  value: number;
-}
-
-interface FlagReportState {
-  label_values: LabelValue[];
-  submittable: boolean;
-}
-
 interface FlaggableElementProps {
   children: React.ReactNode;
   message: Message;
 }
 
-export const FlaggableElement = (props: FlaggableElementProps) => {
-  const [report, updateReport] = useReducer(
-    (state: FlagReportState, action: LoadLabelsAction | UpdateValueAction | ToggleLabelAction): FlagReportState => {
-      const makeState = (label_values: LabelValue[]): FlagReportState => {
-        const submittable = label_values.map(({ checked }) => checked).some(Boolean);
-        return { label_values, submittable };
-      };
-
-      switch (action.type) {
-        case "load_labels":
-          return makeState(
-            action.labels.map((label) => {
-              return { label, checked: false, value: 1 };
-            })
-          );
-        case "toggle_label": {
-          const values_copy = state.label_values.slice();
-          values_copy[action.label_index].checked = action.check;
-          return makeState(values_copy);
-        }
-        case "update_value": {
-          const values_copy = state.label_values.slice();
-          values_copy[action.label_index].value = action.value;
-          return makeState(values_copy);
-        }
-      }
-    },
-    { label_values: [], submittable: false }
-  );
-  const [isEditing, setIsEditing] = useBoolean();
+interface ValidLabelsResponse {
+  valid_labels: Label[];
+}
 
-  const { data, isLoading } = useSWR("/api/valid_labels", get);
-  useEffect(() => {
-    if (isLoading) {
-      return;
-    }
-    if (!data) {
-      updateReport({ type: "load_labels", labels: [] });
-      return;
-    }
-    const { valid_labels } = data;
-    updateReport({ type: "load_labels", labels: valid_labels });
-  }, [data, isLoading]);
+export const FlaggableElement = (props: FlaggableElementProps) => {
+  const { data: response } = useSWRImmutable<ValidLabelsResponse>("/api/valid_labels", get);
+  const { isOpen, onOpen, onClose } = useDisclosure();
+  const { valid_labels } = response || { valid_labels: [] };
+  const [values, setValues] = useState<number[]>([]);
+
+  const submittable =
+    values.some((value) => {
+      return value !== null;
+    }) &&
+    values.length === valid_labels.length &&
+    valid_labels.length > 0;
 
   const { trigger } = useSWRMutation("/api/set_label", post, {
-    onSuccess: setIsEditing.off,
+    onSuccess: onClose,
+    onError: onClose,
   });
 
   const submitResponse = () => {
     const label_map: Map<string, number> = new Map();
-    report.label_values.forEach(({ label, checked, value }) => {
-      if (checked) {
-        label_map.set(label.name, value);
-      }
-    });
+    console.assert(valid_labels.length === values.length);
+    values
+      .map((value, idx) => {
+        return { labelName: valid_labels[idx].name, value };
+      })
```

Comment:
this `map` is redundant, you can get the name from the index in the `forEach` directly.

Reply:
Yeah, not sure how I missed that, I'll blame organic development of the code :laughing:.